Data loss warning on Angular editforms on window.beforeunload

aleixsuau opened the pull request:
https://community.openproject.com/projects/openproject/work_packages/25683

This pull request:

- [x] Warns the user about a possible data loss on the Angular forms (EditFormComponent) when moving out of the  Angular app (ie: clicking on the main side menu items).
- [x] Adds a GlobalEditFormChangesTrackerService that tracks the changes on all the EditFormComponent active in the app.

Note:
This pull request solves the bug but only on the changes made inside and EditFormComponent (ie: wp description). Changes made out of them, like comments on a work package, are not registered on the GlobalEditFormChangesTrackerService and would be lost without any warning when clicking on a link out of the Angular app.

I created a [new work package](https://community.openproject.com/projects/openproject/work_packages/34551) to cover these cases too.

aleixsuau committed:
Data loss warning on Angular editforms on window.beforeunload

aleixsuau committed:
Data loss warning on Angular editforms on window.beforeunload

openprojectci commented on frontend/src/app/modules/fields/edit/edit-form/edit-form.component.ts:
```diff
@@ -44,6 +44,7 @@ import {EditFieldHandler} from "core-app/modules/fields/edit/editing-portal/edit
 import {EditFormRoutingService} from "core-app/modules/fields/edit/edit-form/edit-form-routing.service";
 import {ResourceChangesetCommit} from "core-app/modules/fields/edit/services/hal-resource-editing.service";
+import {GlobalEditFormChangesTrackerService} from "core-app/modules/fields/edit/services/global-edit-form-changes-tracker/global-edit-form-changes-tracker.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/opf/openproject/blob/7a895a2f422ccb1590b4607212148b670ce61d59/frontend/src/app/modules/fields/edit/edit-form/edit-form.component.ts#L48"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

openprojectci replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/opf/openproject/blob/83f0f3ca335b769d053227115d7648f66a062b1d/frontend/src/app/modules/fields/edit/edit-form/edit-form.component.ts#L48"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

aleixsuau committed:
TSLint error fix

aleixsuau committed:
TSLint error fix

aleixsuau committed:
fieldsWithModelChanges typed

aleixsuau committed:
fieldsWithModelChanges typed

openprojectci commented on frontend/src/app/modules/fields/edit/edit-form/edit-form.component.ts:
```diff
@@ -44,6 +44,8 @@ import {EditFieldHandler} from "core-app/modules/fields/edit/editing-portal/edit
 import {ResourceChangesetCommit} from "core-app/modules/fields/edit/services/hal-resource-editing.service";
+import {GlobalEditFormChangesTrackerService} from "core-app/modules/fields/edit/services/global-edit-form-changes-tracker/global-edit-form-changes-tracker.service";
+import {FormattableEditFieldComponent} from "core-app/modules/fields/edit/field-types/formattable-edit-field.component";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/opf/openproject/blob/531e5548faf70fd912e127df04fbdb69ca54c754/frontend/src/app/modules/fields/edit/edit-form/edit-form.component.ts#L49"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

bluedeepimpact committed:
typos, improvements

[ci skip]
typos, improvements (sections about projects)

oliverguenther committed:
Merge remote-tracking branch 'origin/release/11.0' into dev

bluedeepimpact committed:
grammar, typos, wording, additions

[ci skip]

oliverguenther committed:
Improve mysql migration documentation

[ci skip]

bluedeepimpact committed:
extended documentation for Slack plugin

[ci skip]

bsatarnejad committed:
change the teaser of new features in BIM

bsatarnejad committed:
change the teaser of new features in BIM

aleixsuau committed:
Fix: 'Create wp row' created but not shown because of permissions (#8700)

* Fix: 'Create wp row' created but not shown because of permissions

* Removed useless dependencies

oliverguenther committed:
Merge branch 'release/11.0' into dev

ulferts committed:
bump browser

ulferts committed:
bump browser

ulferts committed:
bump aws-sdk-core

ulferts committed:
bump aws-sdk-core

ulferts committed:
bump aws-sdk-s3

ulferts committed:
bump aws-sdk-s3

ulferts committed:
bump crack

ulferts committed:
bump crack

ulferts committed:
bump faker

ulferts committed:
bump faker

ulferts committed:
bump fog-core

ulferts committed:
bump fog-core

ulferts committed:
bump newrelic_rpm

Commit:
update locales from crowdin [ci skip]

Commit:
update locales from crowdin [ci skip]

ulferts committed:
bump parallel_tests

ulferts committed:
bump parallel_tests

ulferts committed:
bump rack-mini-profiler

ulferts committed:
bump rack-mini-profiler

ulferts committed:
bump regexp_parser

ulferts committed:
bump regexp_parser

ulferts committed:
bump rubocop

ulferts committed:
bump rubocop

ulferts committed:
bump icalendar

ulferts committed:
bump icalendar

ulferts committed:
bump nio4r

bsatarnejad committed:
change the text to be the same as standard mode

bsatarnejad committed:
change the text to be the same as standard mode

oliverguenther committed:
Merge branch 'release/11.0' into dev

bsatarnejad committed:
change the link

bsatarnejad committed:
change the link

bluedeepimpact committed:
changed links from blog to docs (for Excel synch)

[ci skip]

ulferts committed:
Merge pull request #8706 from opf/fix/bump_gems

Fix/bump gems

oliverguenther committed:
Merge remote-tracking branch 'origin/release/10.6' into dev

ulferts committed:
Merge branch 'release/11.0' into dev

bsatarnejad committed:
[34358] Unclear warning message displayed when filtered value for assignee board does not match any member (#8695)

* correct the warning message in assignee board when there isn't any member

* make sure that the user type at least a letter

oliverguenther committed:
Merge remote-tracking branch 'origin/release/11.0' into dev

ulferts committed:
always use list cf subquery on list custom field filtering

The check in place before would check if any custom value, when casted to int would not be greater than zero. This also applies to nil -> 0. This then lead to the non list cf subquery being used when filtering. As a lot of custom values have nil for their value attribute, filtering for list custom fields was only possible in very rare cases.

oliverguenther committed:
Merge remote-tracking branch 'origin/release/11.0' into dev

bsatarnejad committed:
Merge pull request #8703 from opf/new-features-teaser-in-bim

[34514] What's new teaser OpenProject BIM 11.0

oliverguenther committed:
Merge branch 'release/11.0' into dev

Commit:
update locales from crowdin [ci skip]

Commit:
update locales from crowdin [ci skip]

crohr committed:
Switch to workflow_dispatch event for Core/CD

ulferts committed:
remove unnecessary documentation

Since having been moved into the core, this separate docs are no longer necessary

ulferts committed:
remove unnecessary documentation

Since having been moved into the core, this separate docs are no longer necessary

ulferts committed:
set project on cost query

ulferts committed:
set project on cost query

ulferts committed:
redirect to project context upon saving

ulferts committed:
redirect to project context upon saving

oliverguenther committed:
Merge branch 'release/11.0' into dev

ulferts committed:
merged reporting_engine into reporting

ulferts committed:
merged reporting_engine into reporting

oliverguenther committed:
Merge remote-tracking branch 'origin/release/11.0' into dev

aleixsuau committed:
GlobalEditFormChangesTrackerService tracking changes only inside the editForm

aleixsuau committed:
GlobalEditFormChangesTrackerService tracking changes only inside the editForm

oliverguenther committed:
Merge remote-tracking branch 'origin/release/11.0' into dev

aleixsuau committed:
Merged with dev

Commit:
update locales from crowdin [ci skip]

Commit:
update locales from crowdin [ci skip]

Commit:
update locales from crowdin [ci skip]

Commit:
update locales from crowdin [ci skip]

ulferts committed:
Merge pull request #8721 from opf/fix/project_scope_of_cost_query

Fix/project scope of cost query

ulferts committed:
Merge branch 'release/11.0' into dev

bluedeepimpact committed:
documentation for template functions

[ci skip]

bluedeepimpact committed:
Merge branch 'dev' of https://github.com/opf/openproject into dev

ulferts committed:
Merge branch 'release/11.0' into dev

aleixsuau committed:
Merge branch 'dev' into fix/25683-warning-when-data-will-be-discarded

aleixsuau committed:
Merge branch 'release/11.0' into fix/25683-warning-when-data-will-be-discarded

aleixsuau committed:
Rebased to release/11.0

aleixsuau commented:
Closing because creating a new branch rebased on release/11.0
